src/language_features/kak_lsp_custom.rs

```diff
@@ -0,0 +1,318 @@
+fn next_or_prev_response(
+    meta: EditorMeta,
+    ctx: &mut Context,
+    maybe_details: Option<(String, KakounePosition, String, SymbolKind)>,
+    hover: bool,
+) {
+    if let Some((filename, symbol_position, name, kind)) = maybe_details {
+        if hover {
+            // Create hover `req_params` just like we do in `hover::text_document_hover()`
+            let req_params = HoverParams {
+                text_document_position_params: TextDocumentPositionParams {
+                    text_document: TextDocumentIdentifier {
+                        uri: Url::from_file_path(&meta.buffile).unwrap(),
+                    },
+                    position: get_lsp_position(&meta.buffile, &symbol_position, ctx).unwrap(),
+                },
+                work_done_progress_params: Default::default(),
+            };
+
+            // This context is shown at the top of the modal
+            let context = format!(
+                                "line {}:{}:{{+b@KindAndName}}{:?} {}{{KindAndName}} (Press `g` to goto this position. Press any other key to continue)",
+                                symbol_position.line, symbol_position.column, kind, escape_brace(&name)
+                            );
+
+            // This kak script is appended after the kakoune `lsp-show-hover`
+            let do_after = format!(
+                "on-key %[eval %sh[
+                     if [ \"$kak_key\" = \"g\" ];
+                     then echo 'info -style modal
+                                exec {}g{}lh
+                                ';
+                     else echo 'info -style modal';
+                     fi
+                 ]]",
+                symbol_position.line, symbol_position.column
+            );
+
+            // Make a HoverRequest to the backend language server and then call `hover::editor_hover()`
+            ctx.call::<HoverRequest, _>(
+                meta,
+                req_params,
+                move |ctx: &mut Context, meta, result| {
+                    editor_hover(
+                        meta,
+                        Some(HoverModal { context, do_after }),
+                        PositionParams {
+                            position: symbol_position,
+                        },
+                        result,
+                        ctx,
+                    )
+                },
+            );
+        } else {
+            let location = format!(
+                "%§{}§ {} {}",
+                filename.replace("§", "§§"),
+                symbol_position.line,
+                symbol_position.column
+            );
+            let command = format!("edit! -existing {}", location);
+            ctx.exec(meta, command);
+        }
+    } else {
+        ctx.exec(
+            meta,
+            "eval %[
+                info -style modal 'Not found!\n\nPress any key to continue'
+                on-key %[info -style modal]
+            ]",
+        );
+    }
+}
```

Comment:
This is really amazing. I've briefly used VSCode, and I liked that it shows the name of the current function at the top of each buffer. So it's not a problem if you forget the name of the current function.

So this was a feature I've definitely been missing, and I like the hover UI a lot. I'll try it some more this week, to see if it stands the test of time. In general I don't use hover much, but this seems very useful.

I'll need some time to fully review this. I have already started writing lots of review comments.
If you don't have time to address them, let me know, then I can keep it to the important bits / take over.

I'm comfortable writing GitHub review comments, because I write them in Kakoune and post them using the GitHub API, but I understand that dealing with lots of comments may be difficult when using the GitHub web interface (Sadly I can't reply to threads using the API just yet). I'm happy to switch to something like email if people prefer that.

Reply:
@krobelus Feel free to add as many review comments as you need to! 

I'll try to address as many as I can this week. For any significant ones that remain, we can discuss what to do about them -- thanks for your offer of taking them up, if necessary.